### PR TITLE
Add warning when calling sparse_coo_tensor._values()

### LINKED ISSUE #70357: torch.sparse_coo_tensor._values() incorrectly returns a detached tensor
### 🐛 Describe the bug

Calling `torch.sparse_coo_tensor._values()` incorrectly detaches the tensor from the autograd graph:

```python
import torch
a = torch.sparse_coo_tensor(
  indices=torch.tensor([[1,2,3]]),
  values=torch.tensor([3., 4., 5.], requires_grad=True),
  size=(1000,)
)

a.requires_grad
# True
a.coalesce().values().requires_grad
# True

# When we call without coalescing, gradients are lost
a._values().requires_grad
# False
```

### Versions

```
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.4 (arm64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.30)
CMake version: version 3.21.4
Libc version: N/A

Python version: 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:25:50)  [Clang 11.1.0 ] (64-bit runtime)
Python platform: macOS-11.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.4
[pip3] pytorch-seed-rl==0.0.1
[pip3] torch==1.10.0
[pip3] torch-geometric==2.0.2
[pip3] torch-scatter==2.0.9
[pip3] torch-sparse==0.6.12
[pip3] torch-spline-conv==1.2.1
[pip3] torchtyping==0.1.4
[pip3] torchvision==0.9.0a0
[conda] numpy                     1.21.4           py38hbf7bb01_0    conda-forge
[conda] pytorch                   1.10.0          cpu_py38ha3d347f_0    conda-forge
[conda] pytorch-seed-rl           0.0.1                    pypi_0    pypi
[conda] torch-geometric           2.0.2                    pypi_0    pypi
[conda] torch-scatter             2.0.9                    pypi_0    pypi
[conda] torch-sparse              0.6.12                   pypi_0    pypi
[conda] torch-spline-conv         1.2.1                    pypi_0    pypi
[conda] torchtyping               0.1.4                    pypi_0    pypi
[conda] torchvision               0.9.1           py38h4ddca67_0_cpu    conda-forge
```

cc @nikitaved @pearu @cpuhrsch @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by smorad**
Fixes https://github.com/pytorch/pytorch/issues/70357

Test:
```python
>>> import torch
>>> torch.sparse_coo_tensor([[]],[])._values()
```

`<stdin>:1: UserWarning: Calling sparse_tensor._values() returns a detached version of the values. If you want to track gradients through the values tensor, call sparse_tensor.coalesce().values() instead. (Triggered internally at  /Users/smorad/code/from_src/pytorch/aten/src/ATen/native/sparse/SparseTensor.cpp:55.)
tensor([])`

### CODE REVIEW

**Comment by smorad** [comment]
Actually, it seems `_values()` is called all over the place in the source code.
```
./aten/src/ATen/native/TensorShape.cpp:          t._values().options().layout_opt(),                                                                                     [91/1974]
./aten/src/ATen/native/TensorShape.cpp:          t._values().options().device_opt(),
./aten/src/ATen/native/TensorShape.cpp:          t._values().options().pinned_memory_opt());
./aten/src/ATen/native/TensorShape.cpp:      vals_pieces.push_back(native::cat({z1, t._values(), z2}, values_dim));
./aten/src/ATen/native/TensorShape.cpp:    new_values = self._values().index_select(0, nzIndices);
./aten/src/ATen/native/TensorShape.cpp:    new_values = self._values().narrow_copy(dense_dim, start, length);
./aten/src/ATen/native/TensorShape.cpp:  auto values = self._values();
./aten/src/ATen/native/TensorShape.cpp:  auto values = self._values();
./aten/src/ATen/native/TensorShape.cpp:  if (self._indices().numel() == 0 && self._values().numel() == 0) {
./aten/src/ATen/native/TensorShape.cpp:    at::sparse::get_sparse_impl(self)->raw_resize_(self._indices().size(0), self._values().dim() - 1, sizes);
./aten/src/ATen/native/TensorShape.cpp:        sparse_dim + 1, dense_dim, sizes, new_indices, self._values(), self.options());
./aten/src/ATen/native/TensorShape.cpp:        sparse_dim, dense_dim + 1, sizes, indices, self._values().unsqueeze(dim - sparse_dim + 1), self.options());
./aten/src/ATen/native/sparse/SparseTensor.cpp:      "Calling sparse_tensor._values() returns a detached version "
./aten/src/ATen/native/sparse/SparseTensor.cpp:  copy_into_sparse(other, self._indices(), self._values(), true);
./aten/src/ATen/native/sparse/SparseTensor.cpp:  copy_into_sparse(self, src._indices(), src._values(), non_blocking);
./aten/src/ATen/native/sparse/SparseTensor.cpp:  Tensor values = self._values().contiguous();
./aten/src/ATen/native/sparse/SparseTensor.cpp:  Tensor mask_values = mask._values();
./aten/src/ATen/native/sparse/SparseTensor.cpp:  Tensor r_values = at::empty(mask_values.sizes(), r._values().options());
./aten/src/ATen/native/sparse/SparseTensor.cpp:    This is a helper function which filter values from `t._values()` using the
./aten/src/ATen/native/sparse/SparseTensor.cpp:  auto t_v = t._values();
./aten/src/ATen/native/sparse/SparseTensor.cpp:  // Step 2: Filter `t._values()` values by matching the flatten `mask_indices`
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:  Tensor values = sparse._values();
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:  Tensor values = sparse._values().to(commonDtype);
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:  Tensor t_values_ = t._values().to(commonDtype);
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:  Tensor s_values_ = src._values().to(commonDtype);
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:    r_values_ = promoted._values().to(r_.scalar_type());
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:    alias_into_sparse(r_, c._indices(), c._values());
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:  Tensor t_values_ = t._values().to(commonDtype);
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:  Tensor s_values_ = src._values().to(commonDtype);
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:  Tensor input_values = input._values();
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:    Tensor grad_values = grad._values();
./aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:  Tensor values =      self_coalesced._values();
./aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cpp:  Tensor mask_values = mask._values();
./aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cpp:  Tensor r_values = at::empty(mask_values.sizes(), r._values().options());
./aten/src/ATen/native/sparse/cuda/SparseMatMul.cu:  Tensor mat1_values = mat1._values().contiguous();
./aten/src/ATen/native/sparse/cuda/SparseMatMul.cu:  Tensor mat2_values = mat2._values().contiguous();
./aten/src/ATen/native/sparse/cuda/SparseMatMul.cu:  auto output_values = result._values();
./aten/src/ATen/native/sparse/cuda/SoftMax.cu:  auto values = input._values().contiguous();
./aten/src/ATen/native/sparse/cuda/SoftMax.cu:  auto out_values = output._values();
./aten/src/ATen/native/sparse/cuda/SoftMax.cu:  auto grad_values = grad._values().contiguous();
./aten/src/ATen/native/sparse/cuda/SoftMax.cu:  auto out_values = output._values().contiguous();
./aten/src/ATen/native/sparse/cuda/SoftMax.cu:  auto values = grad_input._values();
./aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:  Tensor values = self._values();
./aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:    This is a helper function which filter values from `t._values()` using the
./aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:  auto t_values = t._values().contiguous();
./aten/src/ATen/native/sparse/cuda/SparseCUDATensor.cu:  // Step 4: Copy the Filtered `t._values()` using the matches at `t_indices_pos`
./aten/src/ATen/native/sparse/SparseUnaryOps.cpp:  auto values = self._values();
./aten/src/ATen/native/sparse/SparseUnaryOps.cpp:    auto values = self._values();
...
```

This makes it such that even
```python a = torch.sparse_coo_tensor([[]],[]).coalesce()```
prints `<stdin>:1: UserWarning: 0 (Triggered internally at  /Users/smorad/code/from_src/pytorch/aten/src/ATen/native/sparse/SparseTensor.cpp:51.)`

**Review by pearu — COMMENTED** [review]
I have one concern here: internally, various sparse COO algorithms call `_values()` for a reason and there should be no warnings in these cases.

IIUC, the original issue was that users may misuse the internal method and I am not sure that the PR implements the solution at the proper level. I think that if a user misuses a method marked as an internal method, the solution should be implemented at the documentation level: add a note what will be the implications to autograd when calling `._values()`.

**Comment by smorad** [comment]
A solution would be to create a second-layer function wrapper `_values_sparse_py` for `_values_sparse` that is called when calling `tensor._values()` through the python api. This way, internal calls do not need to be changed and will not trigger the warning.

It's not clear to me how the C++/python codegen works, so I don't really know how to do this.

**Comment by pearu** [comment]
One could overload `_values()` in torch/_tensor.py but even at the Python level there exists internal methods that call `_values()`.

I would still recommend revising the approach and implementing a solution at the docs level: if a user chooses to shoot its left leg, others should not be suffering from it.
